contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol

```diff
@@ -410,33 +423,54 @@ abstract contract FunctionsSubscriptions is IFunctionsSubscriptions, IERC677Rece
     emit SubscriptionConsumerAdded(subscriptionId, consumer);
   }
 
-  // @inheritdoc IFunctionsSubscriptions
-  function cancelSubscription(uint64 subscriptionId, address to) external override {
-    _whenNotPaused();
-    _onlySubscriptionOwner(subscriptionId);
-    _onlySenderThatAcceptedToS();
-
-    if (pendingRequestExists(subscriptionId)) {
-      revert CannotRemoveWithPendingRequests();
-    }
-
-    _cancelSubscriptionHelper(subscriptionId, to);
-  }
+  // Overriden in FunctionsRouter.sol
+  function _getSubscriptionDepositDetails() internal virtual returns (uint16, uint72);
 
-  function _cancelSubscriptionHelper(uint64 subscriptionId, address to) private {
+  function _cancelSubscriptionHelper(uint64 subscriptionId, address toAddress, bool checkDepositRefundability) private {
     Subscription memory subscription = s_subscriptions[subscriptionId];
     uint96 balance = subscription.balance;
+    uint64 completedRequests = 0;
+
     // NOTE: loop iterations are bounded by config.maxConsumers
     // If no consumers, does nothing.
     for (uint256 i = 0; i < subscription.consumers.length; ++i) {
-      delete s_consumers[subscription.consumers[i]][subscriptionId];
+      address consumer = subscription.consumers[i];
+      completedRequests += s_consumers[consumer][subscriptionId].completedRequests;
+      delete s_consumers[consumer][subscriptionId];
     }
     delete s_subscriptions[subscriptionId];
-    s_totalLinkBalance -= balance;
 
-    IERC20(i_linkToken).safeTransfer(to, uint256(balance));
+    (uint16 subscriptionDepositMinimumRequests, uint72 subscriptionDepositJuels) = _getSubscriptionDepositDetails();
+
+    IERC20 linkToken = IERC20(i_linkToken);
+
+    // If subscription has not made enough requests, deposit will be forfeited
+    if (completedRequests < subscriptionDepositMinimumRequests && checkDepositRefundability) {
+      uint96 deposit = subscriptionDepositJuels > balance ? balance : subscriptionDepositJuels;
```

Comment:
Open Question: how are we planning to enforce that the minimum subscription balance must be more than the deposit?  Else, a user could just create a subscription with a tiny balance and still abuse S4.

I assume this check is performed in the core node.  However, maybe we should add some sort of minimum funding value to the tooling and docs to prevent user confusion of users not providing enough subscription funds to use S4.

Reply:
It will be enforced in Gateway+core and needs to be documented. Not sure about tooling, probably enough to display an error message.